File: Pinball App 2/Pinball App 2/practice/PracticeIdentityKeying.swift
import Foundation

extension PinballGame {
    var canonicalPracticeKey: String {
        practiceIdentity?.trimmingCharacters(in: .whitespacesAndNewlines).nilIfEmpty ?? slug
    }
}

private let sourceScopedPracticeGameIDPrefix = "source::"

func sourceScopedPracticeGameID(sourceID: String, gameID: String) -> String {
    let normalizedSourceID = canonicalLibrarySourceID(sourceID) ?? sourceID
    return "\(sourceScopedPracticeGameIDPrefix)\(normalizedSourceID)::\(gameID)"
}

func parseSourceScopedPracticeGameID(_ raw: String) -> (sourceID: String?, gameID: String) {
    let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
    guard trimmed.hasPrefix(sourceScopedPracticeGameIDPrefix) else {
        return (nil, trimmed)
    }
    let payload = String(trimmed.dropFirst(sourceScopedPracticeGameIDPrefix.count))
    let parts = payload.components(separatedBy: "::")
    guard parts.count >= 2 else {
        return (nil, trimmed)
    }
    let sourceID = canonicalLibrarySourceID(parts[0]) ?? parts[0]
    let gameID = parts.dropFirst().joined(separator: "::")
    return (sourceID, gameID)
}

func preferredPracticeRepresentative(_ games: [PinballGame], preferredSourceID: String? = nil) -> PinballGame? {
    let candidateGames: [PinballGame]
    if let preferredSourceID = canonicalLibrarySourceID(preferredSourceID) {
        let preferred = games.filter { canonicalLibrarySourceID($0.sourceId) == preferredSourceID }
        candidateGames = preferred.isEmpty ? games : preferred
    } else {
        candidateGames = games
    }

    return candidateGames.max {
        let lhsScore = practiceRepresentativeScore($0)
        let rhsScore = practiceRepresentativeScore($1)
        if lhsScore != rhsScore { return lhsScore < rhsScore }
        return $0.slug.localizedCaseInsensitiveCompare($1.slug) == .orderedDescending
    }
}

private func practiceRepresentativeScore(_ game: PinballGame) -> Int {
    var score = 0
    if canonicalLibrarySourceID(game.sourceId) == gameRoomLibrarySourceID { score += 600 }
    if game.area != nil || game.group != nil || game.pos != nil { score += 260 }
    if let bank = game.bank, bank > 0 { score += 240 }
    if isAvenueLibrarySourceID(game.sourceId) { score += 180 }
    if game.sourceType == .venue { score += 120 }
    if game.name.contains(":") { score += 120 }
    if let variant = game.normalizedVariant?.trimmingCharacters(in: .whitespacesAndNewlines), !variant.isEmpty {
        score += 100
        if variant.localizedCaseInsensitiveContains("anniversary") { score += 120 }
    }
    if game.primaryImageLargeUrl != nil || game.primaryImageUrl != nil { score += 60 }
    if let year = game.year { score += year }
    score += game.name.count
    return score
}

extension PracticeStore {
    private static let practiceIdentityAliases: [String: String] = [:]
    static let practicePreferenceGameIDKeys: [String] = [
        "practice-quick-game-score",
        "practice-quick-game-study",
        "practice-quick-game-practice",
        "practice-quick-game-mechanics",
        "practice-last-viewed-game-id",
        "library-last-viewed-game-id"
    ]
    private var primaryPracticeLookupGamesPool: [PinballGame] {
        let baseGames = allLibraryGames.isEmpty ? games : allLibraryGames
        guard !searchCatalogGames.isEmpty else { return baseGames }
        return baseGames + searchCatalogGames
    }

    private var practiceLookupGamesPool: [PinballGame] {
        let enrichedGames = primaryPracticeLookupGamesPool + bankTemplateGames
        return enrichedGames
    }

    func canonicalPracticeGameID(_ raw: String) -> String {
        let parsed = parseSourceScopedPracticeGameID(raw)
        let trimmed = parsed.gameID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return "" }
        let aliased = Self.practiceIdentityAliases[trimmed] ?? trimmed
        let lookupID = parsed.sourceID.map { sourceScopedPracticeGameID(sourceID: $0, gameID: aliased) } ?? aliased
        if let match = gameForAnyID(lookupID) {
            return match.canonicalPracticeKey
        }
        if let match = legacyPracticeKeyMatch(for: aliased) {
            return match.canonicalPracticeKey
        }
        return aliased
    }

    func gameForAnyID(_ id: String) -> PinballGame? {
        let parsed = parseSourceScopedPracticeGameID(id)
        let trimmed = parsed.gameID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        let pool = parsed.sourceID == nil ? primaryPracticeLookupGamesPool : practiceLookupGamesPool
        if let exactID = pool.first(where: {
            $0.id == trimmed &&
            (parsed.sourceID == nil || canonicalLibrarySourceID($0.sourceId) == parsed.sourceID)
        }) {
            return exactID
        }
        if let exactSlug = pool.first(where: {
            $0.slug == trimmed &&
            (parsed.sourceID == nil || canonicalLibrarySourceID($0.sourceId) == parsed.sourceID)
        }) {
            return exactSlug
        }
        let practiceMatches = pool.filter { $0.canonicalPracticeKey == trimmed }
        if !practiceMatches.isEmpty {
            return preferredPracticeRepresentative(
                practiceMatches,
                preferredSourceID: parsed.sourceID ?? defaultPracticeSourceID
            )
        }
        return nil
    }

    func practiceGamesDeduped() -> [PinballGame] {
        Dictionary(grouping: primaryPracticeLookupGamesPool, by: \.canonicalPracticeKey)
            .compactMap { preferredPracticeRepresentative($0.value, preferredSourceID: defaultPracticeSourceID) }
            .sorted {
                let nameCompare = $0.name.localizedCaseInsensitiveCompare($1.name)
                if nameCompare != .orderedSame { return nameCompare == .orderedAscending }
                return $0.slug.localizedCaseInsensitiveCompare($1.slug) == .orderedAscending
            }
    }

    @discardableResult
    func migratePracticeStateKeysToCanonicalIfNeeded() -> Bool {
        guard !practiceLookupGamesPool.isEmpty else { return false }
        var didMigrateState = false
        if stateNeedsCanonicalMigration(state) {
            state = migratedStateKeysToCanonical(state)
            saveState()
            didMigrateState = true
        }
        migratePracticePreferenceGameIDsToCanonicalIfNeeded()
        return didMigrateState
    }

    private func migratedStateKeysToCanonical(_ current: PracticePersistedState) -> PracticePersistedState {
        func mapID(_ value: String) -> String { canonicalPracticeGameID(value) }

        var next = current
        next.studyEvents = current.studyEvents.map { StudyProgressEvent(id: $0.id, gameID: mapID($0.gameID), task: $0.task, progressPercent: $0.progressPercent, timestamp: $0.timestamp) }
        next.videoProgressEntries = current.videoProgressEntries.map { VideoProgressEntry(id: $0.id, gameID: mapID($0.gameID), kind: $0.kind, value: $0.value, timestamp: $0.timestamp) }
        next.scoreEntries = current.scoreEntries.map { ScoreLogEntry(id: $0.id, gameID: mapID($0.gameID), score: $0.score, context: $0.context, tournamentName: $0.tournamentName, timestamp: $0.timestamp, leagueImported: $0.leagueImported) }
        next.noteEntries = current.noteEntries.map { PracticeNoteEntry(id: $0.id, gameID: mapID($0.gameID), category: $0.category, detail: $0.detail, note: $0.note, timestamp: $0.timestamp) }
        next.journalEntries = current.journalEntries.map {
            JournalEntry(
                id: $0.id,
                gameID: mapID($0.gameID),
                action: $0.action,
                task: $0.task,
                progressPercent: $0.progressPercent,
                videoKind: $0.videoKind,
                videoValue: $0.videoValue,
                score: $0.score,
                scoreContext: $0.scoreContext,
                tournamentName: $0.tournamentName,
                noteCategory: $0.noteCategory,
                noteDetail: $0.noteDetail,
                note: $0.note,
                timestamp: $0.timestamp
            )
        }
        next.customGroups = current.customGroups.map { group in
            var updated = group
            updated.gameIDs = uniqueGameIDsPreservingOrder(group.gameIDs.map(mapID))
            return updated
        }
        next.rulesheetResumeOffsets = Dictionary(uniqueKeysWithValues: current.rulesheetResumeOffsets.map { (mapID($0.key), $0.value) }.filter { !$0.0.isEmpty })
        next.videoResumeHints = Dictionary(uniqueKeysWithValues: current.videoResumeHints.map { (mapID($0.key), $0.value) }.filter { !$0.0.isEmpty })
        next.gameSummaryNotes = Dictionary(uniqueKeysWithValues: current.gameSummaryNotes.map { (mapID($0.key), $0.value) }.filter { !$0.0.isEmpty })
        return next
    }

    private func migratePracticePreferenceGameIDsToCanonicalIfNeeded() {
        let defaults = UserDefaults.standard
        for key in Self.practicePreferenceGameIDKeys {
            let raw = defaults.string(forKey: key) ?? ""
            if raw.isEmpty { continue }
            let canonical = canonicalPracticeGameID(raw)
            if canonical != raw {
                defaults.set(canonical, forKey: key)
            }
        }
    }

    func legacyPracticeKeyMatch(for raw: String) -> PinballGame? {
        let pool = primaryPracticeLookupGamesPool
        if let opdbGroup = extractLikelyOPDBGroupID(from: raw) {
            if let byGroup = pool.first(where: { $0.canonicalPracticeKey.caseInsensitiveCompare(opdbGroup) == .orderedSame }) {
                return byGroup
            }
        }

        let normalized = normalizedLegacyGameKey(raw)
        guard !normalized.isEmpty else { return nil }
        if let slugMatch = pool.first(where: { normalizedLegacyGameKey($0.slug) == normalized }) {
            return slugMatch
        }
        let canonicalMatches = pool.filter { normalizedLegacyGameKey($0.canonicalPracticeKey) == normalized }
        if !canonicalMatches.isEmpty {
            return preferredPracticeRepresentative(canonicalMatches, preferredSourceID: defaultPracticeSourceID)
        }
        return nil
    }

    private func stateNeedsCanonicalMigration(_ current: PracticePersistedState) -> Bool {
        func mapsDiffer(_ value: String) -> Bool {
            let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !trimmed.isEmpty else { return false }
            return canonicalPracticeGameID(trimmed) != trimmed
        }

        if current.studyEvents.contains(where: { mapsDiffer($0.gameID) }) { return true }
        if current.videoProgressEntries.contains(where: { mapsDiffer($0.gameID) }) { return true }
        if current.scoreEntries.contains(where: { mapsDiffer($0.gameID) }) { return true }
        if current.noteEntries.contains(where: { mapsDiffer($0.gameID) }) { return true }
        if current.journalEntries.contains(where: { mapsDiffer($0.gameID) }) { return true }
        if current.customGroups.contains(where: { uniqueGameIDsPreservingOrder($0.gameIDs.map(canonicalPracticeGameID)) != $0.gameIDs }) { return true }
        if current.rulesheetResumeOffsets.keys.contains(where: mapsDiffer) { return true }
        if current.videoResumeHints.keys.contains(where: mapsDiffer) { return true }
        if current.gameSummaryNotes.keys.contains(where: mapsDiffer) { return true }
        return false
    }

    private func extractLikelyOPDBGroupID(from raw: String) -> String? {
        let pattern = #"(?i)\bG[0-9A-Z]{4,}\b"#
        guard let regex = try? NSRegularExpression(pattern: pattern) else { return nil }
        let range = NSRange(raw.startIndex..<raw.endIndex, in: raw)
        guard let match = regex.firstMatch(in: raw, options: [], range: range),
              let tokenRange = Range(match.range, in: raw) else {
            return nil
        }
        return String(raw[tokenRange])
    }

    private func normalizedLegacyGameKey(_ raw: String) -> String {
        raw
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased()
            .replacingOccurrences(of: "&", with: "and")
            .replacingOccurrences(of: #"[^a-z0-9]+"#, with: "-", options: .regularExpression)
            .trimmingCharacters(in: CharacterSet(charactersIn: "-"))
    }
}

private extension String {
    var nilIfEmpty: String? {
        isEmpty ? nil : self
    }
}

File: Pinball App Android/app/src/main/java/com/pillyliu/pinprofandroid/practice/PracticeIdentityKeying.kt
package com.pillyliu.pinprofandroid.practice

import com.pillyliu.pinprofandroid.library.canonicalLibrarySourceId
import com.pillyliu.pinprofandroid.library.isAvenueLibrarySourceId
import com.pillyliu.pinprofandroid.library.isGameRoomLibrarySourceId
import com.pillyliu.pinprofandroid.library.LibrarySource
import com.pillyliu.pinprofandroid.library.LibrarySourceType
import com.pillyliu.pinprofandroid.library.PinballGame
import com.pillyliu.pinprofandroid.library.normalizedVariant
import java.util.Locale

internal val PinballGame.practiceKey: String
    get() = practiceIdentity?.trim()?.takeIf { it.isNotBlank() } ?: slug

private const val SOURCE_SCOPED_PRACTICE_GAME_ID_PREFIX = "source::"

internal fun sourceScopedPracticeGameID(sourceID: String, gameID: String): String {
    val normalizedSourceID = canonicalLibrarySourceId(sourceID) ?: sourceID
    return "$SOURCE_SCOPED_PRACTICE_GAME_ID_PREFIX$normalizedSourceID::$gameID"
}

internal data class ParsedSourceScopedPracticeGameID(
    val sourceID: String?,
    val gameID: String,
)

internal fun parseSourceScopedPracticeGameID(raw: String): ParsedSourceScopedPracticeGameID {
    val trimmed = raw.trim()
    if (!trimmed.startsWith(SOURCE_SCOPED_PRACTICE_GAME_ID_PREFIX)) {
        return ParsedSourceScopedPracticeGameID(sourceID = null, gameID = trimmed)
    }
    val payload = trimmed.removePrefix(SOURCE_SCOPED_PRACTICE_GAME_ID_PREFIX)
    val parts = payload.split("::")
    if (parts.size < 2) {
        return ParsedSourceScopedPracticeGameID(sourceID = null, gameID = trimmed)
    }
    val sourceID = canonicalLibrarySourceId(parts.first()) ?: parts.first()
    val gameID = parts.drop(1).joinToString("::")
    return ParsedSourceScopedPracticeGameID(sourceID = sourceID, gameID = gameID)
}

internal val PinballGame.displayTitleForPractice: String
    get() = name

internal fun PinballGame.matchesPracticeLookupKey(value: String?): Boolean {
    val parsed = parseSourceScopedPracticeGameID(value.orEmpty())
    val key = parsed.gameID.trim()
    if (key.isBlank()) return false
    if (parsed.sourceID != null && canonicalLibrarySourceId(sourceId) != parsed.sourceID) return false
    return key == practiceKey || key == slug
}

internal fun findGameByPracticeLookupKey(games: List<PinballGame>, value: String?): PinballGame? {
    val parsed = parseSourceScopedPracticeGameID(value.orEmpty())
    val key = parsed.gameID.trim()
    if (key.isBlank()) return null
    val exactSlugMatch = games.firstOrNull {
        it.slug == key && (parsed.sourceID == null || canonicalLibrarySourceId(it.sourceId) == parsed.sourceID)
    }
    if (exactSlugMatch != null) return exactSlugMatch

    val practiceMatches = games.filter { it.practiceKey == key }
    if (practiceMatches.isNotEmpty()) {
        return preferredPracticeRepresentative(practiceMatches, preferredSourceId = parsed.sourceID)
    }

    return null
}

internal fun gamesByPracticeLookupKey(games: List<PinballGame>): Map<String, PinballGame> {
    val out = LinkedHashMap<String, PinballGame>()
    games
        .groupBy { it.practiceKey }
        .forEach { (key, grouped) ->
            preferredPracticeRepresentative(grouped)?.let { representative ->
                out[key] = representative
            }
        }
    games.forEach { game ->
        out.putIfAbsent(game.slug, game)
        out[sourceScopedPracticeGameID(game.sourceId, game.practiceKey)] = game
    }
    return out
}

internal fun distinctGamesByPracticeIdentity(games: List<PinballGame>): List<PinballGame> {
    return games
        .groupBy { it.practiceKey }
        .values
        .mapNotNull(::preferredPracticeRepresentative)
        .sortedWith(compareBy<PinballGame> { it.name.lowercase(Locale.US) }.thenBy { it.slug.lowercase(Locale.US) })
}

private val PRACTICE_IDENTITY_ALIASES: Map<String, String> = emptyMap()

internal fun canonicalPracticeKey(value: String?, games: List<PinballGame>): String {
    val parsed = parseSourceScopedPracticeGameID(value.orEmpty())
    val raw = parsed.gameID.trim()
    if (raw.isBlank()) return ""
    val aliased = PRACTICE_IDENTITY_ALIASES[raw] ?: raw
    val resolvedLookup = if (parsed.sourceID != null) {
        sourceScopedPracticeGameID(parsed.sourceID, aliased)
    } else {
        aliased
    }
    return findGameByPracticeLookupKey(games, resolvedLookup)?.practiceKey
        ?: legacyPracticeKeyMatch(games, aliased)?.practiceKey
        ?: aliased
}

internal fun canonicalizeGroupSelectionKey(value: String?, games: List<PinballGame>): String {
    val parsed = parseSourceScopedPracticeGameID(value.orEmpty())
    val resolved = findGameByPracticeLookupKey(games, value)
        ?: legacyPracticeKeyMatch(games, parsed.gameID)
        ?: return canonicalPracticeKey(value, games)
    return parsed.sourceID?.let { sourceScopedPracticeGameID(it, resolved.practiceKey) } ?: resolved.practiceKey
}

internal fun uniqueGroupSelectionIDsPreservingOrder(ids: List<String>, games: List<PinballGame>): List<String> {
    val seen = linkedSetOf<String>()
    val ordered = mutableListOf<String>()
    ids.forEach { id ->
        val normalized = canonicalizeGroupSelectionKey(id, games)
        val canonical = canonicalPracticeKey(normalized, games)
        if (canonical.isBlank() || !seen.add(canonical)) return@forEach
        ordered += normalized
    }
    return ordered
}

internal fun preferredPracticeSelectionKey(
    game: PinballGame,
    selectedSourceId: String?,
    librarySources: List<LibrarySource>,
): String {
    val selectedSource = librarySources.firstOrNull { it.id == selectedSourceId }
    return if (selectedSource?.type == LibrarySourceType.VENUE) {
        sourceScopedPracticeGameID(selectedSource.id, game.practiceKey)
    } else {
        game.practiceKey
    }
}

internal fun migratePracticeStateKeys(state: PracticePersistedState, games: List<PinballGame>): PracticePersistedState {
    fun mapKey(key: String): String = canonicalPracticeKey(key, games)
    fun mapGroupKey(key: String): String = canonicalizeGroupSelectionKey(key, games)

    val migratedGroups = state.groups.map { group ->
        group.copy(gameSlugs = uniqueGroupSelectionIDsPreservingOrder(group.gameSlugs.map(::mapGroupKey), games))
    }
    val migratedScores = state.scores.map { it.copy(gameSlug = mapKey(it.gameSlug)) }
    val migratedNotes = state.notes.map { it.copy(gameSlug = mapKey(it.gameSlug)) }
    val migratedJournal = state.journal.map { it.copy(gameSlug = mapKey(it.gameSlug)) }
    val migratedRulesheetProgress = state.rulesheetProgress
        .mapKeys { (key, _) -> mapKey(key) }
        .filterKeys { it.isNotBlank() }
    val migratedGameSummaryNotes = state.gameSummaryNotes
        .mapKeys { (key, _) -> mapKey(key) }
        .filterKeys { it.isNotBlank() }

    return state.copy(
        groups = migratedGroups,
        scores = migratedScores,
        notes = migratedNotes,
        journal = migratedJournal,
        rulesheetProgress = migratedRulesheetProgress,
        gameSummaryNotes = migratedGameSummaryNotes,
    )
}

internal fun migrateCanonicalPracticeStateKeys(
    state: CanonicalPracticePersistedState,
    games: List<PinballGame>,
): CanonicalPracticePersistedState {
    fun mapKey(key: String): String = canonicalPracticeKey(key, games)
    fun mapGroupKey(key: String): String = canonicalizeGroupSelectionKey(key, games)

    return state.copy(
        studyEvents = state.studyEvents.map { it.copy(gameID = mapKey(it.gameID)) },
        videoProgressEntries = state.videoProgressEntries.map { it.copy(gameID = mapKey(it.gameID)) },
        scoreEntries = state.scoreEntries.map { it.copy(gameID = mapKey(it.gameID)) },
        noteEntries = state.noteEntries.map { it.copy(gameID = mapKey(it.gameID)) },
        journalEntries = state.journalEntries.map { it.copy(gameID = mapKey(it.gameID)) },
        customGroups = state.customGroups.map { group ->
            group.copy(gameIDs = uniqueGroupSelectionIDsPreservingOrder(group.gameIDs.map(::mapGroupKey), games))
        },
        rulesheetResumeOffsets = state.rulesheetResumeOffsets
            .mapKeys { (key, _) -> mapKey(key) }
            .filterKeys { it.isNotBlank() },
        videoResumeHints = state.videoResumeHints
            .mapKeys { (key, _) -> mapKey(key) }
            .filterKeys { it.isNotBlank() },
        gameSummaryNotes = state.gameSummaryNotes
            .mapKeys { (key, _) -> mapKey(key) }
            .filterKeys { it.isNotBlank() },
    )
}

internal fun legacyPracticeKeyMatch(games: List<PinballGame>, raw: String): PinballGame? {
    extractLikelyOpdbGroup(raw)?.let { token ->
        games.firstOrNull { it.practiceKey.equals(token, ignoreCase = true) }?.let { return it }
    }
    val normalized = normalizedLegacyGameKey(raw)
    if (normalized.isBlank()) return null
    return games.firstOrNull { normalizedLegacyGameKey(it.slug) == normalized }
        ?: games.firstOrNull { normalizedLegacyGameKey(it.practiceKey) == normalized }
}

private fun extractLikelyOpdbGroup(raw: String): String? {
    val match = Regex("\\bG[0-9A-Za-z]{4,}\\b", RegexOption.IGNORE_CASE).find(raw) ?: return null
    return match.value
}

private fun normalizedLegacyGameKey(raw: String): String =
    raw.trim()
        .lowercase(Locale.US)
        .replace("&", "and")
        .replace(Regex("[^a-z0-9]+"), "-")
        .trim('-')

private fun preferredPracticeRepresentative(
    games: List<PinballGame>,
    preferredSourceId: String? = null,
): PinballGame? {
    val candidateGames = preferredSourceId
        ?.let { preferred ->
            val normalizedPreferred = canonicalLibrarySourceId(preferred) ?: preferred
            games.filter { canonicalLibrarySourceId(it.sourceId) == normalizedPreferred }
                .ifEmpty { games }
        }
        ?: games
    return candidateGames.maxWithOrNull(compareBy<PinballGame>(::practiceRepresentativeScore).thenBy { it.slug })
}

private fun practiceRepresentativeScore(game: PinballGame): Int {
    var score = 0
    if (isGameRoomLibrarySourceId(game.sourceId)) score += 600
    if (game.area != null || game.group != null || game.position != null) score += 260
    if ((game.bank ?: 0) > 0) score += 240
    if (isAvenueLibrarySourceId(game.sourceId)) score += 180
    if (game.sourceType.name == "VENUE") score += 120
    if (game.name.contains(":")) score += 120
    if (!game.normalizedVariant.isNullOrBlank()) score += 100
    if (game.normalizedVariant?.contains("anniversary", ignoreCase = true) == true) score += 120
    if (!game.primaryImageLargeUrl.isNullOrBlank() || !game.primaryImageUrl.isNullOrBlank()) score += 60
    if (game.year != null) score += game.year
    score += game.name.length
    return score
}
